ServiceFabric.PubSubActors/SubscriberServices/SubscriberStatelessServiceBase.cs

```diff
@@ -132,11 +137,62 @@ protected void ServiceEventSourceMessage(string message, [CallerMemberName] stri
             ServiceEventSourceMessageCallback?.Invoke($"{caller} - {message}");
         }
 
-        private class SubscriptionDefinition
+        /// <summary>
+        /// Scans this service for attributes of type <see cref="SubscribeAttribute"/> and corresponding methods that can handle 
+        /// messages. 
+        /// </summary>
+        internal void DiscoverHandlers()
+        {
+            Type taskType = typeof(Task);
+            var methods = GetType().GetMethods(BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public);
+            var handlesAttributes = GetType().GetCustomAttributes(typeof(SubscribeAttribute), false).Cast<SubscribeAttribute>();
+            foreach (var attribute in handlesAttributes)
+            {
+                var target = methods.SingleOrDefault(m =>
+                {
+                    var parameters = m.GetParameters();
+                    var argument = parameters.FirstOrDefault();
+                    return parameters.Length == 1 
+                           && m.ReturnType == taskType 
+                           && argument != null 
+                           && argument.ParameterType == attribute.MessageType;
+                });
+                if (target == null)
+                {
+                    throw new InvalidOperationException($"Service {GetType().FullName} is marked with a SubscribeAttribute for MessageType {attribute.MessageType}, but it does not have an async method that handles it.");
+                }
+
+                Subscriptions[attribute.MessageType.FullName] = new SubscriptionDefinition
+                {
+                    MessageType = attribute.MessageType,
+                    Handler = m => (Task)target.Invoke(this, new[]{m})
+                };
+            }
+        }
+    }
+
+    public class SubscriptionDefinition
+    {
+        public Uri Broker { get; set; }
+        public Type MessageType { get; set; }
+        public Func<object, Task> Handler { get; set; }
+    }
+
+    /// <summary>
+    /// Marks a service as being capable of receiving messages.
+    /// Follows convention that class has method with signature 'Task MethodName(MessageType message)'
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class, AllowMultiple = true, Inherited = false)]
+    public class SubscribeAttribute : Attribute
+    {
+        /// <summary>
+        /// Type of message.
+        /// </summary>
+        public Type MessageType { get; }
```

Comment:
We should also add a parameter for the broker Uri.

Reply:
Agreed, made some changes accordingly.
Also, I'm thinking we could add a bootstrapper to the broker service, that would detect subscriber services and automatically register them. Potentially, we can avoid introduction of a required service base class, making the library more applicable in projects. Still struggling with how to discover the service CLR type while potentially running inside a different application. 

Maybe we just add a `Register` operation to `ISubscriberService` and invoke that?

Reply:
The bootstrapper is an interesting idea.  If the Broker can automatically discover and register subscribers, it could dramatically simplify the consumer.  Being able to work with a CLR type that is not present in the application is a tough problem.  I'm not sure it's possible.  I was not able to put much time into this today, but I will see if I can figure something out tomorrow.